src/overlays/actors/ovl_En_River_Sound/z_en_river_sound.c

```diff
@@ -27,15 +30,212 @@ const ActorInit En_River_Sound_InitVars = {
+    Vec3f sp2C;
+    Vec3f sp20;
+    Vec3f sp14;
+    f32 f;
+    f32 g;
+    f32 h;
+
+    sp20.x = arg0->x - arg2->x;
+    sp20.y = arg0->y - arg2->y;
+    sp20.z = arg0->z - arg2->z;
+
+    sp14.x = arg1->x - arg2->x;
+    sp14.y = arg1->y - arg2->y;
+    sp14.z = arg1->z - arg2->z;
+
+    sp2C.x = sp14.x - sp20.x;
+    sp2C.y = sp14.y - sp20.y;
+    sp2C.z = sp14.z - sp20.z;
 
+    f = sp20.z * sp2C.z + (sp2C.x * sp20.x + sp2C.y * sp20.y);
+    g = sp14.z * sp2C.z + (sp2C.x * sp14.x + sp2C.y * sp14.y);
+    h = sp2C.z * sp2C.z + (sp2C.x * sp2C.x + sp2C.y * sp2C.y);
+
+    if (g * f < 0) {
+        arg3->x = sp2C.x * (-f / h) + arg0->x;
+        arg3->y = sp2C.y * (-f / h) + arg0->y;
+        arg3->z = sp2C.z * (-f / h) + arg0->z;
+
+        return 1;
+    } else {
+        return 0;
+    }
+}
+#else
 #pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_River_Sound/func_80AE6A54.s")
+#endif
+
+s32 func_80AE6BC0(Vec3s* points, s32 numPoints, Vec3f* pos, Vec3f* res) {
+    s32 i;
+    s32 pointIdx;
+    s32 sp78[2] = { 0, 0 };
+    Vec3f pointLoc;
+    Vec3f sp60;
+    Vec3f sp54;
+    Vec3f vec;
+    f32 pointDist = 10000.f;
+    Vec3s* point;
+
+    for (i = 0; i < numPoints; i++) {
+        f32 d;
+        vec.x = points[i].x;
+        vec.y = points[i].y;
+        vec.z = points[i].z;
+        d = Math_Vec3f_DistXYZ(pos, &vec);
+        if (d < pointDist) {
+            pointDist = d;
+            pointIdx = i;
+        }
+    }
+    if (pointDist >= 10000.f)
+        return 0;
+
+    point = &points[pointIdx];
+    pointLoc.x = point->x;
+    pointLoc.y = point->y;
+    pointLoc.z = point->z;
+    if (pointIdx != 0) {
+        vec.x = point[-1].x;
+        vec.y = point[-1].y;
+        vec.z = point[-1].z;
+        sp78[0] = func_80AE6A54(&vec, &pointLoc, pos, &sp54);
+    }
+    if (pointIdx + 1 != numPoints) {
+        vec.x = point[1].x;
+        vec.y = point[1].y;
+        vec.z = point[1].z;
+        sp78[1] = func_80AE6A54(&pointLoc, &vec, pos, &sp60);
+    }
+
+    if (sp78[0] && sp78[1]) {
+        if (!func_80AE6A54(&sp54, &sp60, pos, res)) {
+            res->x = (sp54.x + sp60.x) * .5f;
+            res->y = (sp54.y + sp60.y) * .5f;
+            res->z = (sp54.z + sp60.z) * .5f;
+        }
+    } else if (sp78[0]) {
+        res->x = sp54.x;
+        res->y = sp54.y;
+        res->z = sp54.z;
+    } else if (sp78[1]) {
+        res->x = sp60.x;
+        res->y = sp60.y;
+        res->z = sp60.z;
+    } else {
+        res->x = pointLoc.x;
+        res->y = pointLoc.y;
+        res->z = pointLoc.z;
+    }
+
+    return 1;
+}
+
+void EnRiverSound_Update(Actor* thisx, GlobalContext* globalCtx) {
+    Path* path;
+    Vec3f* pos;
+    Player* player = PLAYER;
+    EnRiverSound* this = THIS;
+    s32 sp34;
+
+    if (this->actor.params == 0 || this->actor.params == 4 || this->actor.params == 5) {
+        path = &globalCtx->setupPathList[this->pathIndex];
+        pos = &thisx->world.pos;
```

Comment:
This seems to be the only place where `thisx` is required.

Reply:
So am I understanding this right, use `thisx` for all accesses to the actor base, and `this` for accesses that are specific to `EnRiverSound`?

Reply:
thats right. our current approach is: try to be consistent with using only `this`. however if `thisx` is required in one place, its enough to prove that the dev who wrote this preferred to use `thisx` for base accesses instead. so you can do it for all in that function